File: 202308/2616.go
import "sort"

func minimizeMax(nums []int, p int) int {
	sort.Ints(nums)
	n := len(nums)
	lo, hi := 0, nums[n-1]

	ok := func(v int) bool {
		cnt := 0
		for i := 1; i < n; i++ {
			if nums[i]-nums[i-1] <= v {
				cnt++
				i++
			}
		}
		if cnt >= p {
			return true
		}
		return false
	}

	ans := -1
	for lo <= hi {
		mid := lo + (hi-lo)/2
		if ok(mid) {
			ans = mid
			hi = mid - 1
		} else {
			lo = mid + 1
		}
	}
	return ans
}

File: 202506/2616.py
class Solution:
    def minimizeMax(self, nums: List[int], p: int) -> int:
        n = len(nums)
        nums.sort()

        def ok(u):
            i = 1
            cnt = 0
            while i < n:
                if nums[i] - nums[i - 1] <= u:
                    cnt += 1
                    i += 2
                else:
                    i += 1
            return cnt >= p

        ans = -1
        lo, hi = 0, max(nums) - min(nums)
        while lo <= hi:
            mi = lo + hi >> 1
            if ok(mi):
                ans = mi
                hi = mi - 1
            else:
                lo = mi + 1
        return ans
